libuavcan/src/transport/uc_dispatcher.cpp

```diff
@@ -149,41 +149,73 @@ void Dispatcher::ListenerRegistry::handleFrame(const RxFrame& frame)
 void Dispatcher::handleFrame(const CanRxFrame& can_frame)
 {
     RxFrame frame;
-    if (!frame.parse(can_frame))
-    {
-        // This is not counted as a transport error
-        UAVCAN_TRACE("Dispatcher", "Invalid CAN frame received: %s", can_frame.toString().c_str());
-        return;
-    }
 
-    if ((frame.getDstNodeID() != NodeID::Broadcast) &&
-        (frame.getDstNodeID() != getNodeID()))
-    {
-        return;
-    }
+    const Protocol iface_protocol = getCanIOManager().getIfaceProtocol(can_frame.iface_index);
 
-    switch (frame.getTransferType())
+    // next block of operations only valid for UAVCAN, which requires extended frame
+    if (can_frame.isExtended() && iface_protocol == Protocol::Standard)
     {
-    case TransferTypeMessageBroadcast:
-    {
-        lmsg_.handleFrame(frame);
-        break;
-    }
-    case TransferTypeServiceRequest:
-    {
-        lsrv_req_.handleFrame(frame);
-        break;
+        if (!frame.parse(can_frame))
+        {
+            // This is not counted as a transport error
+            UAVCAN_TRACE("Dispatcher", "Invalid CAN frame received: %s", can_frame.toString().c_str());
+            return;
+        }
+
+        if ((frame.getDstNodeID() != NodeID::Broadcast) &&
+            (frame.getDstNodeID() != getNodeID()))
+        {
+            return;
+        }
     }
-    case TransferTypeServiceResponse:
+
+    if (iface_protocol == Protocol::Standard && can_frame.isExtended())
```

Comment:
I'm thinking if we should change this IF block to be clearer what the purpose is: if the protocol expectation is set to "custom", then iterate the custom handlers. If the custom protocols should only apply for not-extended frames, only iterate them in that case. Make the latter a second configuration option. In all other cases feed the standard listeners.

```
if ( custom && (!custom_11bit || !can_frame.isExtended()) ) {
    // custom handlers
} else {
    // standard handlers
}
```